spikewrap/data_classes/base.py

```diff
@@ -0,0 +1,138 @@
+import fnmatch
+from collections import UserDict
+from collections.abc import ItemsView, KeysView, ValuesView
+from dataclasses import dataclass
+from itertools import chain
+from pathlib import Path
+from typing import Callable, Dict, List, Literal
+
+
+@dataclass
+class BaseUserDict(UserDict):
+    """
+    Base class for `PreprocessingData` and `SortingData`
+    used for checking and formatting `base_path`, `sub_name`
+    and `run_names`. The layout of the `rawdata` and
+    `derivatives` folder is identical up to the run
+    folder, allowing use of this class for
+    preprocessing and sorting.
+
+    Base UserDict that implements the
+    keys(), values() and items() convenience functions."""
+
+    base_path: Path
+    sub_name: str
+    sessions_and_runs: Dict
+
+    def __post_init__(self) -> None:
+        self.data: Dict = {}
+        self.base_path = Path(self.base_path)
+        self.check_run_names_are_formatted_as_list()
+
+    def check_run_names_are_formatted_as_list(self) -> None:
+        """"""
+        for key, value in self.sessions_and_runs.items():
+            if not isinstance(value, List):
+                assert isinstance(
+                    value, str
+                ), "Run names must be string or list of strings"
+                self.sessions_and_runs[key] = [value]
+
+    def preprocessing_sessions_and_runs(self):  # TODO: type hint
```

Comment:
My inderstanding here is that you want to be sure that the keys in the nested dict structure (`data`) will be ordered. Is that correct?

Reply:
This unecessarily complex function has been refactored in #121 (it took me ages to figure out what it was actually doing 😅)